index.rst

```diff
@@ -13,17 +13,256 @@
+   Datastore can be expected to make this operation atomic, either because it is naturally atomic for its storage backing or via writing a temporary and moving it.
+   We do have to accept the possibility of failures leaving partially-written temporary files around.
+   When the Datastore is backed by storage that requires signed URLs, this write is to a staging area and uses non-signed URIs.
+
+4. If the butler has a Registry, either held directly (as in a "full butler" today) or as a client of a butler server, call ``Registry.insertDatasets`` with both the ``DatasetRef`` and the records returned by the Datastore.
+   Database transactions can be used to ensure that all tables in the Registry (including those for datastore records) are updated consistently or left unchanged.
+   If this operation fails, or the butler does not have a Registry (e.g. ``QuantumBackedButler``), the dataset is left in a valid state: it is in the Datastore, but not the Registry.
+   If the Datastore is backed by storage that requires signed URLs, this call also tells the server to transfer the artifacts from the staging area to permanent storage after checking for permissions.
+   This must happen before the database changes are committed.
+
+This has two major advantages over our current ``put`` implementation:
+
+- there is no database transaction over the Datastore write, keeping transactions small and reducing contention for database connections;
+
+- for a client/server butler, there is no alternation between object-store and http operations, just one and then the other, reducing latency (assuming the data ID has indeed been obtained in advance), and increasing the possiblity that the client Datastore can just be a regular ``FileDatastore``.
+  Any database transactions needed can also happen entirely in a single server operation.
+
+This change will make it so Datastore sees conflicting writes before Registry.
+While these are expected to be rare, and thus conflict resolution does not need to be highly optimized, is important that it does not have the potential to corrupt the repository.
+The easiest way to do this is to configure Datastore storage backends to prohibit writes to locations where files already exist; this is the default behavior for most backends of interest, but we currently explicitly permit silent clobbering to work around problems with automatic retries in BPS.
+:cite:`DMTN-205` outlines an alternative approach to automatric retries that is better for provenance and should be adopted instead.
+
+``Butler.import`` and ``Butler.transfer_from``
+""""""""""""""""""""""""""""""""""""""""""""""
+
+These operations would behave like vectorized versions of ``Butler.put``, with all Datastore writes (if nontrivial transfers are required) occurring before a single Registry or butler server operation that (within a transaction) adds datasets and the associated datastore records to the database.
```

Comment:
The main issue with transfer_from is the large numbers of records that are involved for 250k transfers in batch. On the plus side those are presumably not using client server and we will just have to have a records registry API that allows it to do what it does now inside Datastore which is to return 10k of records at a time to keep things manageable. We can defer all the dimension insertion and dataset insertion until after we have done the file transfers or do the dimensions up front still as you say below.